File: ex3/src/GraphAlgo.py
import json
import queue
from math import inf
import random
from typing import List
import numpy as np
import matplotlib.pyplot as plt
from NodeData import node_data
from DiGraph import DiGraph
from GraphAlgoInterface import GraphAlgoInterface
from GraphInterface import GraphInterface


class GraphAlgo(GraphAlgoInterface):
    """
    This class represents set of algorithms on a directed weighted graph, and some important actions on graph,including:
    0. shortestPath(int src, int dest);
    1. save_to_json(file); // save graph as JSON file
    2. load_from_json(file); // load graph from aJSON file
    3. connected_component(node_id) //Returns the connected component of a node
    4.connected_components //Returns all connected components of the graph
    @author Shlomo GLick
    """

    def __init__(self, my_graph: DiGraph = None):
        if (my_graph == None):
            self.graph_algo = DiGraph()
        else:
            self.graph_algo = my_graph

    def get_graph(self) -> GraphInterface:
        """
        This method return the underlying graph of which this class works
        :return:GraphInterface
        """
        return self.graph_algo

    def load_from_json(self, file_name: str) -> bool:
        """
        This method load a graph to this graph algorithm.
        if the file was successfully loaded - the underlying graph
        of this class will be changed (to the loaded one), in case the
        graph was not loaded the original graph should remain "as is".
        :param file_name:  - file name of JSON file
        :return: True - iff the graph was successfully loaded.
        """
        try:
            with open(file_name, "r") as file:
                self.graph_algo = self.dict_to_graph(json.load(file))
                file.closed
                return True

        except IOError as e:
            print(e)

        return False

    def save_to_json(self, file_name: str) -> bool:
        """
        This method saves the graph to a given file name in Json format
        :param file_name: - the file name (may include a relative path).
        :return: True if and only if the file was successfuly saved
        """
        if self.get_graph() == None: return False
        try:
            with open(file_name, "w") as file:
                json.dump(self.as_dict(), indent=4, fp=file)
                return True
        except IOError as e:
            print(e)

        return False

    def shortest_path(self, id1: int, id2: int) -> (float, list):
        """
        This method calculates and returns a list of node_data representing the shortest path Between two nodes
         and the weight of the path

        :param id1: - start node of the traversal
        :param id2: - end (target) node
        :return:tuple(weight: float ,path: list)
        """
        null = (float(inf), [])
        graph = self.get_graph()
        if graph == None: return null
        if not id1 in graph.get_all_v() or not id2 in graph.get_all_v(): return null
        fathers = self.init_tag(id1)
        dis = graph.get_all_v()[id2].get_tag()
        if dis == -1: return null
        path = self.get_path(id1, id2, fathers, [])
        graph.get_all_v()[id2].reset_tag()
        path.reverse()
        ans = (dis, path)
        return ans

    def connected_component(self, id1: int) -> list:
        """
        Finds the Strongly Connected Component(SCC) that node id1 is a part of.
        :param id1:  : The node id
        :return: The list of nodes in the SCC
        If the graph is None or id1 is not in the graph, return an empty list []
        """
        ans = []
        graph = self.get_graph()
        if graph == None: return ans
        if not id1 in graph.get_all_v(): return ans
        self.init_tag(id1)
        for node in graph.get_all_v().values():
            if node.get_tag() != -1:
                ans.append(node.get_key())
        revers_my_graph = self.revers_graph(graph)
        self.__init__(revers_my_graph)
        self.init_tag(id1)
        del_from_ans = []
        for i in ans:
            node = revers_my_graph.get_all_v()[i]
            if node.get_tag() == -1:
                del_from_ans.append(i)
        for i in del_from_ans:
            ans.remove(i)
        node_data(0).reset_tag()
        self.__init__(graph)
        return ans

    def connected_components(self) -> List[list]:
        """
        Finds all the Strongly Connected Component(SCC) in the graph.
        :return:The list all SCC
        If the graph is None return an empty list []
        """
        ans = []
        if self.get_graph() == None: return ans
        for node_id in self.get_graph().get_all_v():
            flag = False
            for group in ans:
                if node_id in group:
                    flag = True
                    break
            if not flag:
                ans.append(self.connected_component(node_id))

        return ans

    def plot_graph(self) -> None:
        """
        Plots the graph.
        If the nodes have a position, the nodes will be placed there.
        Otherwise, they will be placed in a random but elegant manner.
        Note: The display size is adjusted to the graph it displays
        :return:None
        """
        rnd = random
        graph = self.get_graph()
        if graph == None or graph.node_size == 0: return
        pos_all = {}
        list_x, list_y = [], []
        for i in graph.get_all_v():
            node = graph.get_node(i)
            key = node.get_key()
            pos = node.get_location()
            pos_all[key] = pos
            if pos == (0, 0, 0):
                rnd.seed(key)
                x = rnd.random() + np.sin(key)
                y = rnd.random() + np.cos(key)
            else:
                x = pos[0]
                y = pos[1]
            pos_all[key] = (x, y, 0)
            list_x.append(x)
            list_y.append(y)
        min_x = min(list_x)
        max_x = max(list_x)
        min_y = min(list_y)
        max_y = max(list_y)
        dx = (max_x - min_x)
        dy = (max_y - min_y)
        for i in pos_all:
            x = list_x[i]
            y = list_y[i]
            plt.text(x, y + (dy / 40), f"{i}", color="b")
        plt.plot(list_x, list_y, 'ro')
        for key in graph.get_all_v():
            pos_current = pos_all[key]
            edges = graph.all_out_edges_of_node(key)
            for i in edges:
                pos_nei = pos_all[i]
                dis_x = pos_nei[0] - pos_current[0]
                dis_y = pos_nei[1] - pos_current[1]
                plt.arrow(pos_current[0], pos_current[1], dis_x, dis_y, color="gray", length_includes_head=True,
                          head_width=dx / 80, head_length=dx / 66.666, width=dx / 10000, fc='k', ec='k')

        extension_x = dx / 20
        extension_y = dy / 20
        plt.axis([min_x - extension_x, max_x + extension_x, min_y - extension_y, max_y + extension_y])
        plt.title("my graph:")
        plt.show()

    def get_path(self, src: int, dest: int, fathers: {}, path: []) -> list:
        """
        Runs from the dest and passes each time from a vertex to his father with the help of a fathers dictionary
        and adds them to the "path" list until he reaches src
        where he stops
        :param src:
        :param dest:
        :param fathers:
        :param path:
        :return:a ptah list from dest to src
        """
        path.append(dest)
        if (dest == src):
            return path
        else:
            return self.get_path(src, fathers[dest], fathers, path)

    def init_tag(self, src: int = -1) -> dict:
        """
        Updates the tag to all vertices
        so that for each vertex the tag is equal to the lowest weight of a trajectory to the current vertex
        The algorithm:
        Let's start with the first vertex whose distance from itself is 0
        We will run on all its neighbors
        and update them the tag that will be equal to the tag of node  the  current + the weight of the side that connects them.
        We will then put all of his neighbors in the priority queue (lowest weight).
        And run until the line is over
        So in each iteration we will perform the same action on the vertex at the top of the priority queue
        when we go over its neighbors and for that only in case the tag value has not yet been updated or the new value is lower than the old one
        we will perform the tag update and re-insert queue
        In addition to the function there is also a dictionary of fathers
        where every time we update the tag
        we also add to it the vertex from which it came to the ancestral list
        :param src:
        :return: fathers: dict
        """

        graph=self.get_graph()
        my_priority_queue = queue.PriorityQueue()
        fathers = {src: -1}
        node_src = graph.get_all_v()[src]
        node_src.set_tag(0)
        my_priority_queue.put(node_src)
        while len(my_priority_queue.queue) != 0:
            node_current = my_priority_queue.get()
            key_current = node_current.get_key()
            tag_current = node_current.get_tag()
            for key_dest,w in graph.all_out_edges_of_node(key_current).items():
                distance = tag_current + w
                node_dest = graph.get_all_v()[key_dest]
                tag_dest = node_dest.get_tag()
                if tag_dest > distance or tag_dest == -1:
                    node_dest.set_tag(distance)
                    fathers[key_dest] = key_current
                    my_priority_queue.put(node_dest)

        return fathers

    def as_dict(self) -> dict:
        """
        :return: the dictionary that represents the graph
        """
        graph=self.get_graph()
        if  graph== None: return None
        Edges = []
        nodes = graph.get_all_v()
        for node_id in nodes:
            for id,w in  graph.all_out_edges_of_node(node_id).items():
                src = node_id
                dest = id
                edge_dic = {"src": src, "w": w, "dest": dest}
                Edges.append(edge_dic)
        Nodes = []
        nodes2 = self.graph_algo.get_all_v()
        for i in nodes2:
            node = nodes2[i]
            node_id = node.key
            pos = f"{node.pos[0]},{node.pos[1]},{node.pos[2]}"
            n_dict = {"pos": pos, "id": node_id}
            Nodes.append(n_dict)
        return {"Edges": Edges, "Nodes": Nodes}

    def dict_to_graph(self, dict_graph) -> DiGraph:
        """
        Gets a dictionary that represents the graph and converts it to a graph
        :param dict_graph:
        :return: graph DiGraph
        """
        graph = DiGraph()
        Edges = dict_graph["Edges"]
        Nodes = dict_graph["Nodes"]
        for node in Nodes:
            if "pos" in node:
                str_pos = node["pos"]
                pos_tup = tuple(map(float, str_pos.split(',')))
                graph.add_node(node["id"], pos_tup)
            else:
                graph.add_node(node["id"])
        for edge in Edges:
            graph.add_edge(edge["src"], edge["dest"], edge["w"])
        return graph

    def revers_graph(self, graph: DiGraph) -> DiGraph:
        """
        Get a graph and return the same graph only with inverted ribs
        :param graph:
        :return: Inverted graph
        """
        ans = DiGraph()
        for i in graph.get_all_v():
            ans.add_node(i)
        for node_current in graph.get_all_v():
            for neighbor,w in graph.all_in_edges_of_node(node_current).items():
                ans.add_edge(node_current, neighbor, w)
        return ans








"""

        node = node_data(0)  # creating an object from this class
        node.get_key()  # getting the id of the node
        node.get_tag()  # getting the tag of the node
        node.get_location()  # getting the geometric location of the node
        node.set_tag(3)  # setting the tag of the node
        node.set_location((4.5, 1, 2))  # setting the location of the node
        node.reset_tag()  # Resets the tags of all vertices in the class

        graph = DiGraph()               # creating graph object
        graph.add_node(0)               # adding a node(0) to the graph
        graph.add_node(1)               # adding a node(1) to the graph
        graph.add_edge(0,1,0.5)         # creat an edge between node(0) to node(1) with weight=0.5
        graph.get_all_v()               # dictionary of all the nodes in the Graph -> {0: node(0), 1: node(1)}
        graph.all_out_edges_of_node(0)  # getting all the nodes connected from node_id(0) -> {1: 0.5}
        graph.all_in_edges_of_node(1)   # getting all the nodes connected to (into) node(1) -> {0: 0.5}
        graph.v_size()                  # getting the number of nodes in the graph -> 2
        graph.e_size()                  # getting the number of edges in the graph -> 1
        graph.get_mc()                  # getting the Mode Count -> 3
        graph.remove_edge(0,1)          # deletes the edge(0,1) from the graph
        graph.remove_node(1)            # Deletes the node(1) from the graph

        algo =GraphAlgo(graph)           # creating graph algo object and initialization a graph
        algo.get_graph()                 # getting the graph -> graph
        algo.save_to_json("file_name")   # saving the graph to a file - JSON format
        algo.load_from_json("file_name") # loading a graph from a file - JSON format
        algo.shortest_path(0,100)   # getting  list of node_data representing the shortest path Between two nodes and the weight of the path
        algo.connected_component(0)      # getting Strongly Connected Component(SCC) that node id1 is a part of -> [0, 1]
        algo.connected_components()      #  getting all the Strongly Connected Component(SCC) in the graph.
        algo.plot_graph()                # Draws the graph on a graphical window


"""

File: ex0/Graph_Algo.java
package ex0;


import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Graph_Algo implements graph_algorithms {
    private graph myGraphAlgo = new Graph_DS();
    private int connectedSize = 0;

    public Graph_Algo() {
        this.myGraphAlgo = null;
    }


    @Override
    public void init(graph g) {
        this.myGraphAlgo = g;
    }

    /**
     * Creates a new graph
     * Runs on all nodes in the old graph
     * And for each node creates a new node with the same data
     * And adds it to the new graph
     * Then goes over all its neighbors from the old graph
     * And if they have already been copied to the new graph adds them to the neighbors list of the new node
     *
     * @return g
     */
    @Override
    public graph copy() {   //    Complications  ->  O(1)*n*deg(Vi)=total deg(g)*O(1)=2edgeSize*O(1)=2k*O(1)=    O(k)   (So that k = edgeSize)
        if (this.myGraphAlgo==null) return null;
        Graph_DS g = new Graph_DS();
        for (node_data oldNode : this.myGraphAlgo.getV()) {
            NodeData newNode = new NodeData(oldNode.getKey(), oldNode.getInfo());
            newNode.setTag(oldNode.getTag());
            g.addNode(newNode);
            for (node_data neighbor : oldNode.getNi()) {
                if (g.getNode(neighbor.getKey()) != null) {
                    g.connect(neighbor.getKey(), newNode.getKey());
                }
            }
        }
        g.setMC(this.myGraphAlgo.getMC());
        return g;
    }



    /**
     * run the initConnectedSize function
     * That will update connectedSize
     * To the number of nodes connected to the first node.
     * Then if the graph is connected
     * We will get the number of nodes connected to the first node
     * Equal to the number of nodes in the graph
     *
     * @return
     */
    @Override
    public boolean isConnected() {                                   // O(e) e=edgeSize
        if (this.myGraphAlgo.nodeSize() == 0) return true;
        initConnectedSizeAndTag();
        boolean ask = (connectedSize == this.myGraphAlgo.nodeSize());
        connectedSize = 0;
        NodeData.reset();                                                //Reset the tag at all nodes in the graph for the next test
        return ask;
    }

    /**
     * The function is initialized at each node the tag:
     * Tag = distance to the first junction
     * And initializes connectedSize:
     * connectedSize = The number of members that managed to reach from the first including the first
     */
    private void initConnectedSizeAndTag() {      // O(e)
        node_data first = getFirstNode(this.myGraphAlgo);
        initConnectedSizeAndTag(first.getKey());
    }

    /**
     * @param g
     * @return The first node in the graph g
     */
    private node_data getFirstNode(graph g) {
        Iterator it = this.myGraphAlgo.getV().iterator();
        return (node_data) it.next();
    }


    /**
     * We will put the "current" node into the queue
     * The function is initialized at each node the tag:
     * Tag = distance to the "current" junction
     * And initializes connectedSize:
     * connectedSize = The number of members that managed to reach from the "current" including the "current"
     */

    private void initConnectedSizeAndTag(int current) {  // O(e)
        Queue<node_data> q1 = new LinkedList<node_data>();
        Queue<node_data> q2 = new LinkedList<node_data>();
        node_data first = this.myGraphAlgo.getNode(current);
        first.setTag(0);
        connectedSize++;
        q1.add(first);
        initConnectedSizeAndTag(q1, q2);           // O(e)
    }


    /**
     * Gets two queues Q1, Q2
     * So that in Q1 there is an "current" junction
     * The function is initialized at each node the tag:
     * Tag = distance to the "current" junction
     * And the function is initialized in "connectedSize"
     * The number of nodes connected to "current" including "current"
     *
     * @param q1
     * @param q2
     */
    /*
    Method:
    We will go through all the nodes by distance levels from the first node
    At each vertex at the distance level i we will go over all its neighbors and update the tag to
            (i + 1)
    So also we will know that we have already been there because
            (Tag! = -1)
    And we will add them to "Q2"
    When in "Q1" the nodes are at a distance i from the first node and in "Q2"
    Distance (i + 1)
    Then we will reactivate the function and switch between the queues
    And so we get that each round represents a distance level from the first node.
    And we go all the way.
    When we reach a state where both queues are empty it means that we have passed all the nodes at distance i
    And no nodes were added at a distance (i + 1)
    So we went through all the connected components
    */
    private void initConnectedSizeAndTag(Queue<node_data> q1, Queue<node_data> q2) {   //Complications: O(e) e = edgeSize
        while (!q1.isEmpty()) {
            node_data current = q1.poll();
            int TagNew = current.getTag() + 1;
            for (node_data node : current.getNi()) {
                if (node.getTag() == -1) {
                    node.setTag(TagNew);
                    q2.add(node);
                    connectedSize++;
                }
            }
        }
        if (q2.isEmpty())
            return;                                             //There are no nodes at a distance (i + 1) The function will end
        initConnectedSizeAndTag(q2, q1);
        return;
    }


    /**
     * We will update the "tag" of each node = node distance from the node with key = src
     * We return the tag of "dest"
     * because it is equal to the distance of "dest"  From src
     */

    @Override
    public int shortestPathDist(int src, int dest) {
        initConnectedSizeAndTag(src);
        int tag = this.myGraphAlgo.getNode(dest).getTag();
        NodeData.reset();
        return tag;
    }

    /**
     * We will update the "tag" of each node
     * Tag = node distance from the node with key dest
     * We do this using the initConnectedSizeAndTag(dest) function
     * Then
     * In case the "tag" of node.get (src) = -1
     * This means that the nodes are not connected
     * Therefore there is no trajectory in their sons
     * and we will return null
     * Otherwise we will create a Linklist and add the first node to it
     * Then using the AddNodesToTheList function we will add all the other nodes of the route to the list
     * Then to finish
     * We will reset the tag of all nodes using the NodeData.reset function
     * And return the received list
     *
     * @param src  - start node
     * @param dest - end (target) node
     * @return
     */

    @Override
    public List<node_data> shortestPath(int src, int dest) {
        initConnectedSizeAndTag(dest);
        if (this.myGraphAlgo.getNode(src).getTag() == -1) return null;
        List<node_data> ask = new LinkedList<node_data>();
        ask.add(myGraphAlgo.getNode(src));
        AddNodesToTheList(ask, myGraphAlgo.getNode(src));
        NodeData.reset();
        return ask;
    }

    /**
     * We will start with "src"
     * and then look for its neighbor with a smaller "Tag"
     * and then that means it is closer to dest
     * we will move to it we will add it to LinkedList and do exactly the same action
     * only this time closer to dest in step 1 so we repeat the action until we find "dest" in neighbors collection
     * We will put it in the list
     * And we will stop
     *
     * @param ask
     * @param current
     */
    private void AddNodesToTheList(List<node_data> ask, node_data current) {
        if (current.getTag() == 0) {
            return;
        }
        for (node_data node : current.getNi()) {
            if (node.getTag() < current.getTag() && node.getTag() != -1) {
                ask.add(node);
                AddNodesToTheList(ask, node);
                break;
            }
        }
        return;
    }
}






